From: Sven Roederer <devel-sven@geroedel.de>
Date: Sun, 14 Nov 2021 02:35:36 +0100
Subject: mt76: Squashed all commits to backport updates from master

This are just cherry-picked:
    (cherry picked from commit 94c41ef2ef3e443e21ac099d4f38b790db4d9ccf)
    (cherry picked from commit 66cbf5fd4e9c0eb33234ee0378cd1036a18e00fe)
    (cherry picked from commit 192c41001a46ed2c4545aa82e4271802e3357f04)
    (cherry picked from commit ade56b8d9e81553315ba9feaef2e207ad55b06f5)
    (cherry picked from commit 9d994f35b4acc5243209d837b0730f9543ed533e)
    (cherry picked from commit 4f2fd0215b4a125d84ccedde6d6fa323e56bd404)
    (cherry picked from commit 479a2a90f782209b29dd9746d241fc1e67836309)
    (cherry picked from commit 99a22d48f2003910f0968369b61b9131a8f53d80)

on TPLink WR841-v13 I've seen this error quite soon causing the Wifi to break

[325800.532829] ------------[ cut here ]------------
[325800.537659] WARNING: CPU: 0 PID: 18747 at target-mipsel_24kc_musl/linux-ramips_mt76x8/mt76-2021-06-06-22b69033/mt7603/mac.c:206 0x834e44bc [mt7603e@29bc07c4+0x9480]
[325800.552659] Modules linked in: batman_adv mt7603e mt76 mac80211 iptable_nat ipt_REJECT cfg80211 xt_time xt_tcpudp xt_tcpmss xt_statistic xt_state xt_recent xt_nat xt_multiport xt_mark xt_mac xt_limit xt_length xt_hl xt_helper xt_ecn xt_dscp xt_conntrack xt_connmark xt_connlimit xt_connbytes xt_comment xt_TCPMSS xt_REDIRECT xt_MASQUERADE xt_LOG xt_HL xt_FLOWOFFLOAD xt_DSCP xt_CT xt_CLASSIFY nf_reject_ipv4 nf_nat nf_log_ipv4 nf_flow_table_hw nf_flow_table nf_conncount libcrc32c iptable_raw iptable_mangle iptable_filter ipt_ECN ip_tables compat act_connmark nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 sch_tbf sch_ingress sch_htb sch_hfsc em_u32 cls_u32 cls_tcindex cls_route cls_matchall cls_fw cls_flow cls_basic act_skbedit act_mirred act_gact nf_log_ipv6 nf_log_common ip6table_mangle ip6table_filter ip6_tables ip6t_REJECT x_tables nf_reject_ipv6 ifb ipip tunnel4 ip_tunnel veth leds_gpio gpio_button_hotplug crc16 crc32c_generic crypto_hash
[325800.636868] CPU: 0 PID: 18747 Comm: kworker/u3:2 Not tainted 5.4.154 #0
[325800.643673] Workqueue: napi_workq 0x8032d9b4
[325800.648098] Stack : 80560000 82232400 804d9a80 834e44bc 00000000 00000000 00000000 00000000
[325800.656675]         00000000 00000000 00000000 00000000 00000000 00000001 8236fbb8 a839da39
[325800.665245]         8236fc50 00000000 00000000 00000000 00000038 804bf0e4 6e203a65 5f697061
[325800.673815]         00000000 00003f51 00000000 78302071 00000000 8236fb98 00000000 834e44bc
[325800.682385]         00000009 00000002 00000002 834f0000 00000000 8029a278 00000000 806b0000
[325800.690956]         ...
[325800.693527] Call Trace:
[325800.693545] [<834e44bc>] 0x834e44bc [mt7603e@29bc07c4+0x9480]
[325800.702029] [<804bf0e4>] 0x804bf0e4
[325800.705669] [<834e44bc>] 0x834e44bc [mt7603e@29bc07c4+0x9480]
[325800.711620] [<8029a278>] 0x8029a278
[325800.715254] [<8000973c>] 0x8000973c
[325800.718884] [<80009744>] 0x80009744
[325800.722518] [<800237b8>] 0x800237b8
[325800.726157] [<834e44bc>] 0x834e44bc [mt7603e@29bc07c4+0x9480]
[325800.732085] [<80023860>] 0x80023860
[325800.735758] [<834e44bc>] 0x834e44bc [mt7603e@29bc07c4+0x9480]
[325800.741690] [<835b8324>] 0x835b8324 [mac80211@cfbf6458+0x7c310]
[325800.747800] [<80046cdc>] 0x80046cdc
[325800.751475] [<834e4654>] 0x834e4654 [mt7603e@29bc07c4+0x9480]
[325800.757404] [<834c5998>] 0x834c5998 [mt76@3ef57c9e+0x9440]
[325800.763068] [<834c3550>] 0x834c3550 [mt76@3ef57c9e+0x9440]
[325800.768752] [<834e1a18>] 0x834e1a18 [mt7603e@29bc07c4+0x9480]
[325800.774693] [<834c47e8>] 0x834c47e8 [mt76@3ef57c9e+0x9440]
[325800.780410] [<834c153c>] 0x834c153c [mt76@3ef57c9e+0x9440]
[325800.786129] [<8032d8e4>] 0x8032d8e4
[325800.789776] [<8032d9d8>] 0x8032d9d8
[325800.793422] [<8032d9f4>] 0x8032d9f4
[325800.797055] [<804c01ec>] 0x804c01ec
[325800.800701] [<8003a4f4>] 0x8003a4f4
[325800.804388] [<8003a8b4>] 0x8003a8b4
[325800.808021] [<8003a74c>] 0x8003a74c
[325800.811655] [<804c0424>] 0x804c0424
[325800.815300] [<8003a74c>] 0x8003a74c
[325800.818931] [<8003f560>] 0x8003f560
[325800.822565] [<8003f428>] 0x8003f428
[325800.826211] [<8003f428>] 0x8003f428
[325800.829846] [<80005078>] 0x80005078
[325800.833477]
[325800.835083] ---[ end trace 332d71ac9f6091f1 ]---

diff --git a/package/kernel/mt76/Makefile b/package/kernel/mt76/Makefile
index e4051d8347dcfb08b02c95a6ed2ab6f40b9b2a31..3bb95f9a5c227db1a46cfc22593c701dd915db24 100644
--- a/package/kernel/mt76/Makefile
+++ b/package/kernel/mt76/Makefile
@@ -8,9 +8,9 @@ PKG_LICENSE_FILES:=
 
 PKG_SOURCE_URL:=https://github.com/openwrt/mt76
 PKG_SOURCE_PROTO:=git
-PKG_SOURCE_DATE:=2021-06-06
-PKG_SOURCE_VERSION:=22b690334c0f49b11534cc2e331c9d5e17c4a0bc
-PKG_MIRROR_HASH:=ff5e563935919d2e40c1e7254ef3bc06f7ecc5e69f8ddd12903e8f5de942d630
+PKG_SOURCE_DATE:=2021-10-23.1
+PKG_SOURCE_VERSION:=f6bde7ba82eebfbcba952bc436c5ab680fe754f1
+PKG_MIRROR_HASH:=2790686b4c1320950bf653c1fe23814f133b37971a31fc79cc88b352999c1ea4
 
 PKG_MAINTAINER:=Felix Fietkau <nbd@nbd.name>
 PKG_BUILD_PARALLEL:=1
@@ -151,6 +151,14 @@ define KernelPackage/mt76-connac
   FILES:= $(PKG_BUILD_DIR)/mt76-connac-lib.ko
 endef
 
+define KernelPackage/mt76-sdio
+  $(KernelPackage/mt76-default)
+  TITLE:=MediaTek MT7615/MT79xx SDIO driver common code
+  HIDDEN:=1
+  DEPENDS+=+kmod-mt76-core +kmod-mmc
+  FILES:= $(PKG_BUILD_DIR)/mt76-sdio.ko
+endef
+
 define KernelPackage/mt7615-common
   $(KernelPackage/mt76-default)
   TITLE:=MediaTek MT7615 wireless driver common code
@@ -195,9 +203,8 @@ endef
 define KernelPackage/mt7663s
   $(KernelPackage/mt76-default)
   TITLE:=MediaTek MT7663s wireless driver
-  DEPENDS+=+kmod-mmc +kmod-mt7615-common +kmod-mt7663-usb-sdio
+  DEPENDS+=+kmod-mt76-sdio +kmod-mt7615-common +kmod-mt7663-usb-sdio
   FILES:= \
-	$(PKG_BUILD_DIR)/mt76-sdio.ko \
 	$(PKG_BUILD_DIR)/mt7615/mt7663s.ko
   AUTOLOAD:=$(call AutoProbe,mt7663s)
 endef
@@ -218,10 +225,25 @@ define KernelPackage/mt7915e
   AUTOLOAD:=$(call AutoProbe,mt7915e)
 endef
 
+define KernelPackage/mt7921-common
+  TITLE:=MediaTek MT7615 wireless driver common code
+  HIDDEN:=1
+  DEPENDS+=@PCI_SUPPORT +kmod-mt76-core +kmod-mt76-connac
+  FILES:= $(PKG_BUILD_DIR)/mt7921/mt7921-common.ko
+endef
+
+define KernelPackage/mt7921s
+  $(KernelPackage/mt76-default)
+  TITLE:=MediaTek MT7921s wireless driver
+  DEPENDS+=@PCI_SUPPORT +kmod-mt76-connac +kmod-mt76-sdio +kmod-mt7921-common
+  FILES:= $(PKG_BUILD_DIR)/mt7921/mt7921s.ko
+  AUTOLOAD:=$(call AutoProbe,mt7921s)
+endef
+
 define KernelPackage/mt7921e
   $(KernelPackage/mt76-default)
   TITLE:=MediaTek MT7921e wireless driver
-  DEPENDS+=@PCI_SUPPORT +kmod-mt76-connac
+  DEPENDS+=@PCI_SUPPORT +kmod-mt76-connac +kmod-mt7921-common
   FILES:= $(PKG_BUILD_DIR)/mt7921/mt7921e.ko
   AUTOLOAD:=$(call AutoProbe,mt7921e)
 endef
@@ -287,6 +309,9 @@ endif
 ifdef CONFIG_PACKAGE_kmod-mt76-connac
   PKG_MAKE_FLAGS += CONFIG_MT76_CONNAC_LIB=m
 endif
+ifdef CONFIG_PACKAGE_kmod-mt76-sdio
+  PKG_MAKE_FLAGS += CONFIG_MT76_SDIO=m
+endif
 ifdef CONFIG_PACKAGE_kmod-mt7615-common
   PKG_MAKE_FLAGS += CONFIG_MT7615_COMMON=m
 endif
@@ -301,7 +326,6 @@ ifdef CONFIG_PACKAGE_kmod-mt7663-usb-sdio
   PKG_MAKE_FLAGS += CONFIG_MT7663_USB_SDIO_COMMON=m
 endif
 ifdef CONFIG_PACKAGE_kmod-mt7663s
-  PKG_MAKE_FLAGS += CONFIG_MT76_SDIO=m
   PKG_MAKE_FLAGS += CONFIG_MT7663S=m
 endif
 ifdef CONFIG_PACKAGE_kmod-mt7663u
@@ -310,6 +334,12 @@ endif
 ifdef CONFIG_PACKAGE_kmod-mt7915e
   PKG_MAKE_FLAGS += CONFIG_MT7915E=m
 endif
+ifdef CONFIG_PACKAGE_kmod-mt7921-common
+  PKG_MAKE_FLAGS += CONFIG_MT7921_COMMON=m
+endif
+ifdef CONFIG_PACKAGE_kmod-mt7921s
+  PKG_MAKE_FLAGS += CONFIG_MT7921S=m
+endif
 ifdef CONFIG_PACKAGE_kmod-mt7921e
   PKG_MAKE_FLAGS += CONFIG_MT7921E=m
 endif
@@ -432,6 +462,7 @@ $(eval $(call KernelPackage,mt76x2u))
 $(eval $(call KernelPackage,mt76x2))
 $(eval $(call KernelPackage,mt7603))
 $(eval $(call KernelPackage,mt76-connac))
+$(eval $(call KernelPackage,mt76-sdio))
 $(eval $(call KernelPackage,mt7615-common))
 $(eval $(call KernelPackage,mt7615-firmware))
 $(eval $(call KernelPackage,mt7615e))
@@ -441,6 +472,8 @@ $(eval $(call KernelPackage,mt7663-usb-sdio))
 $(eval $(call KernelPackage,mt7663u))
 $(eval $(call KernelPackage,mt7663s))
 $(eval $(call KernelPackage,mt7915e))
+$(eval $(call KernelPackage,mt7921-common))
+$(eval $(call KernelPackage,mt7921s))
 $(eval $(call KernelPackage,mt7921e))
 $(eval $(call KernelPackage,mt76))
 $(eval $(call BuildPackage,mt76-test))
